---
title: "Homework 1 - answers"
author: "Almog Angel"
date: "01/12/2024"
output: html_document
---

In the following tasks you will use statistics and sequence alignment to analyze real COVID-19 world data.

- Load packages
```{r}
library(tidyverse)

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Biostrings")

library(Biostrings)

BiocManager::install("msa")

library(msa)
library(msaR)
library(seqinr)
library(ape)
```

- Set your working directory
```{r}
# Enter your code here:
path <- getwd()
setwd(path)
```

## Task 1 - Data cleaning

- Read "COVID19_line_list_data.csv" into a variable called "covid_data"
```{r}
# Enter your code here:
covid_data <- read.csv("COVID19_line_list_data.csv")
```

- Take a look at the data
```{r}
View(covid_data)
```

- Find a way to remove empty columns:
```{r}
# Enter your code here:
covid_data %>% 
          discard(~all(is.na(.) | . =="")) -> covid_data

```

- Find a way to deal with missing values (NAs) in  "age" and "gender" columns
```{r}
# Enter your code here:
covid_data %>% 
           drop_na(.,c(age,gender)) -> covid_data
                    
```

- Assuming 'death' is a binary variable (0 for no, 1 for yes) that represent categorical data
- Convert 'death' to a factor class that contains either "yes" or "no"
```{r}
# Enter your code here:
covid_data$death  %>%
                  factor(., levels = c(0,1), labels = c("no", "yes"))-> covid_data$death

covid_data$death[is.na(covid_data$death)] <- "yes"
```


## Task 2 - Descriptive statistics analysis 

- Print a descriptive summary of "age"
```{r}
# Enter your code here:
summary(covid_data$age)
```

- Distribution:
- (1) Generate histogram for "age" distribution and color the bars by "death"
* Use: position = 'identity', bins = 50, alpha = 0.5
```{r}
# Enter your code here:
covid_data %>% 
          ggplot(., aes(x=age, color=death)) +
          geom_histogram(position = 'identity', bins = 50, alpha = 0.5)

```

- (2) Generate bar plot that visualize the number of dead and alive per gender
```{r}
# Enter your code here:
covid_data %>% 
          ggplot(., aes(x=gender, fill=death)) + 
          geom_bar()

tasd=table(covid_data$gender, covid_data$death)
prop.table(tasd, margin = 1)
```


## Task 3 - Statistical Testing


- Run Chi-squared test to investigate the association between gender and death
```{r}
# Enter your code here:
chisq.test(x=covid_data$gender, y=covid_data$death)
```

- Why do we use Chi-squared and not t-test?
```{}
Type your answer here:
the two variables are categorical rather than numerical.
```

- Define the Null Hypothesis of the Chi-squared Test in this example
```{}
Type your answer here:
H_0: gender has no effect on mortality due to covid.
```

- What is the p-value? Is it significant (alpha = 0.05)? Write your conclusions from the statistical test results.
```{}
Type your answer here:
p-value is the probability to get correlation at least as big as we got, given H_0 is true.
the p-value is 0.0057, smaller than alpha = 0.05, thus we reject H_0.
in other words - there is a statistical basis to suspect gender has an effect on morality.
```


According to the media, older people are more likely to die from COVID-19.
- Test this claim statistically
- Frame your null and alternative hypothesis here:
```{}
My null hypothesis:
H_0: age has no correlation with mortality due to covid.
My alternative hypothesis:
age has a correlation on mortality due to covid.
```

- Check if the assumption of normality is valid for your data (use alpha of 0.05)
```{r}
# Enter your code here:
shapiro.test(covid_data$age)
```

- Run a statistical test:
- If you assume normality -> use t-test 
- else -> use Wilcoxon rank-sum test 
```{r}
# Enter your code here:
wilcox.test(covid_data$age[covid_data$death == "no"], covid_data$age[covid_data$death == "yes"])
```

- Report the p-value. Is it significant? Write your conclusions from the statistical test results.
```{}
Type your answer here:
p-value = 2.6e-16, it is very small and hence very significant.
we can safely reject H_0 and claim that the alternative hypothesis holds:
age has a correlation with mortality due to covid.
```

- Read "owid-covid-data.csv" from "https://covid.ourworldindata.org/data/owid-covid-data.csv" into a variable called "covid_world"
```{r}
# Enter your code here:
covid_world <- read.csv('https://covid.ourworldindata.org/data/owid-covid-data.csv')

```

- Take a look at the data
```{r}
View(covid_world)
```


Your friend suggest that countries with high Human Development Index (HDI) have a greater risk to die from COVID-19

- Make a second table and call it "covid_hdi_vs_deaths":
  (1) Select rows with "date" equal to "2024-01-21"
  (2) Select the following columns: "location", "human_development_index", "population", "total_deaths"
  (3) Remove rows with NAs
  (4) Calculate the total deaths per million people to a column named "total_deaths_per_million" (round the number)
```{r}
# Enter your code here:
 covid_world %>% 
            subset(., date == "2024-01-21") %>%
            select(.,location, human_development_index, population, total_deaths) %>%
            drop_na(.,) -> covid_hdi_vs_deaths

covid_hdi_vs_deaths$total_deaths_per_million <- round(1e6*covid_hdi_vs_deaths$total_deaths/covid_hdi_vs_deaths$population)
```

- Use a scatterplot to visualize the relations between HDI and the total deaths per million
```{r}
# Enter your code here:
plot(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$total_deaths_per_million)
```

- Calculate the Pearson and Spearman correlation between HDI and the total deaths per million
```{r}
# Enter your code here:
cor.test(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$total_deaths_per_million, method = "pearson")
cor.test(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$total_deaths_per_million, method = "spearman")

```

- Report the Spearman correlation coefficient, the strength of the correlation (weak/moderate/strong/very strong) and the direction (negative/positive)
```{}
Type your answer here:
the Spearman correlation coefficient is 0.6, it is a strong positive correlation.
```

- Do you agree with the claim that increased in HDI causes high COVID-19 mortality? (write in detail)
```{}
Type your answer here:
statistically speaking, citizens of higher HDI countries are more likely to die from COVID-19.
but, statsits can only describe correlation and not causation,
we cannot infer a causative relation between higher HDI and COVID mortality.
domain knowladge can be used to match the statsics with a causative mechanism, which might involve additional variables.
```

- Find another feature in "covid_world" that explain the correlation between HDI and COVID-19 mortality
* Hint: you already have the answer in the first part of the task
```{r}
# Enter your code here:
 covid_world %>% 
            subset(., date == "2024-01-21") %>%
            select(.,location, human_development_index, population, total_deaths, median_age) %>%
            drop_na(.,) -> covid_hdi_vs_deaths

covid_hdi_vs_deaths$total_deaths_per_million <- round(1e6*covid_hdi_vs_deaths$total_deaths/covid_hdi_vs_deaths$population)

plot(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$median_age)
plot(covid_hdi_vs_deaths$median_age, covid_hdi_vs_deaths$total_deaths_per_million)
plot(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$total_deaths_per_million)

cor.test(covid_hdi_vs_deaths$human_development_index, covid_hdi_vs_deaths$median_age, method = "pearson")
cor.test(covid_hdi_vs_deaths$median_age, covid_hdi_vs_deaths$total_deaths_per_million, method = "pearson")
```

```{}
Write your explanation here: 
we know that COVID-19 is more dangerous to older people.
we can see that HDI is very strongly correlated with median age, and that
median age is strongly correlated with mortality.
thus we can claim that the older population in higher HDI countries is more 
vulnurable to COVID.
```


## Task 4 - Sequence Alignment

- Take a look at the "Biostrings" package vignettes
```{r}
browseVignettes("Biostrings")
```

- Read the file "covid_spike_variants.fasta"
This file contain the amino acids sequence of the COVID-19 spike protein from different variants

- Read the file using the correct function from the package "Biostrings" and assign to a variable called "variants"
```{r}
# Enter your code here:
variants <- readAAStringSet("covid_spike_variants.fasta")
variants
```

- How many amino acids are in the Alpha variant?
```{}
Type your answer here:
1270
```

- Read the documentation for the Multiple Sequence Alignment (msa) function from the package "msa"
```{r}
?msa()
```

- Run MSA for the sequences in "variants" and assign the results to "variants.msa"
```{r}
# Enter your code here:
variants.msa <- msa(variants, type = "protein")
```

- Take a look at the results using the package "msaR" (notice: you can scroll right and left to see all the sequence)
```{r}
msaR(AAMultipleAlignment(variants.msa), colorscheme = "taylor")
```

- (1) Which amino-acids appear in position #13? (type "AMINO_ACID_CODE" in the console)
- (2) Write an example of a SNP that can cause the change in amino acid as we see in the Epsilon variant
```{}
Type your answer here:

```

Phylogenetic tree for the COVID19 variants:
- Use the package "seqinr" to generate a distance matrix from the MSA results and save it in a variable called "distMat"
```{r}
# Enter your code here:

```

- Use the package "ape" to apply "neighbor-joining" (nj) clustering algorithm to construct a phylogenetic tree
- Save the results in a variable called "ptree"
```{r}
# Enter your code here:

```

- Look at the phylogenetic tree and answer the following questions:
```{r}
ape::plot.phylo(ptree)
```

- (1) Which variant is the closest to Gamma?
- (2) Which pair seems to emerged from the Alpha variant?
      (a) Epsilon and Delta
      (b) Beta and Gamma
      (c) Epsilon and Delta
      
```{}
Type your answer here:

```

- Save this R Markdown as HTML before submitting
(1) Go to "Knit" > "Knit to HTML"
(2) Change the file name to your IDs (i.e, "123456789_123456789.html")